README.md

# Express Load

The _express-load_ module provides the ability to load scripts into an Express instance from specified directories or files. 
Make large express MVC applications easier to develop by allowing a logical file separation without having to require your scripts. 
See the examples folder for information.

Express Load can be used to autoload models, routes, schemas, configs, controllers, object maps... etc...

You get access to the autoloaded files in the Express application instance to keep out of the global namespace. 
This also allows access to the scripts via the request object. `req.app`

A script at `controllers/user.js` becomes available as `app.controllers.user` or `req.app.controllers.user` in a request.

## Installation

	$ npm install express-load

## Usage

```js
var load = require('express-load');

load('config')
  .then('routes')
  .into(app);
```

## Simple Express Load Example

### controllers/site.js

```js
exports.index = function(req, res, next) {
  res.send('Hello world!');
};
```

### routes/site.js

```js
module.exports = function(app) {

  var site = app.controllers.site;
  
  app.get('/',
    site.index
  );

};
```

### app.js

```js
var express = require('express')
  , load = require('express-load');

var app = express();

load('controllers')
  .then('routes')
  .into(app);

app.listen(3000)
```

## Load Order

The basic load order is the order that is specified in code, for example, you will want to load models before controllers and controllers before routes:

```js
load('models')
  .then('controllers')
  .then('routes')
  .into(app);
```

This would simply load all models, all controllers, all routes in that order (alphabetical order within the directories).

### Specifying Complex Order

If you have two files in the foo folder _a.js_ and _z.js_ and you need _z.js_ loaded BEFORE _a.js_ you would simply do the following:

```js
load('foo/z.js').then('foo').into(app);
```

_express-load_ will recognise the order and will not add it again later down the chain.

## Logging

Logging is off by default but can be enabled in the following way:

```js
load('controllers', {verbose:true}).into(app);
```

See the verbose example in the examples folder.

## Files and folders

_express-load_ will ignore hidden files and folders (by leading period) unless you explicitly define them to be loaded.

_express-load_ will by default only load files ending with `.js`, `.node`, `.coffee`, `.sjs`, `.json` extensions, unless you set checkext option to false : no file extension check at all. You can also change the list of allowed extension (in extlist option).

```js
load('controllers', {checkext:true, extlist:['.js','.myextension']}).into(app);
```

### Nested Folders

If you had nested folders like the following example: 

	models
		humans
			cool.js
			not.js
		animals
			dog.js
			cat.js

You would end up with the scripts being available in the following structure:

```js
app.models.humans.cool
app.models.humans.not
app.models.animals.dog
app.models.animals.cat
```

### Getting the Express Application instance

This can be done in one of two ways:

#### In a script that is auto-loaded...

```js
module.exports = function(app) {
  console.log(app);
};
```

A script will only be loaded with parameters if module.exports is a function. Multiple parameters can be passed to the script in the `into` method, for example:

```js
load('controllers').into(app, parameter, another);
```

#### From a request object...

```js
app.get('/', function(req, res, next) {
  console.log(req.app);
});
```

### File and Folder names vs Object namespace examples

The names of files and folders are used to create the namespace therefore rules have been put in place. dashes `-` and periods `.` are removed and the following character is uppercased. For example:

	some-controllers/my.controller.js -> *.someControllers.myController

Please see the examples folder for working examples of _express-load_ in action.

# License 

(The MIT License)

Copyright (c) 2012 Jarrad Seers &lt;jarrad@jarradseers.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


HISTORY.md
1.1.7 / 2013-02-02
==================

	* External logger support
	* Updated package.json

1.1.2 / 2012-09-16
==================

	* Allow multiple parameters for into method
	* Updated Readme
	* Updated package.json

1.1.1 / 2012-09-15
==================

	* Fixed logging
	* Fixed recursive hidden directories
	* Updated package.json

1.1.0 / 2012-08-23
==================

	* Updated Exress dev dependencies
	* Added verbose option, logging is off by default
	* Updated Readme
	* Updated package.json

1.0.0 / 2012-08-09
==================

	* Major re-write, now allows chaining
	* Improved examples
	* Added MVC example
	* Updated package.json
	* Updated Readme

0.0.4 / 2012-08-09
==================

	* Added autoConfigure feature
	* Cleaned up indentation and comments
	* Improved configuration example
	* Updated package.json
	* Updated Readme

0.0.3 / 2012-08-08
==================

	* Ignore directories
	* Fixed simple example
	* Updated package.json

0.0.2 / 2012-08-07
==================

	* Updated Readme
	* Updated License
	* Updated package.json

0.0.1 / 2012-08-07
==================

	* Initial release


lib/express-load.js
/*!
 *  ExpressLoad.
 *  Autoload config, models, controllers etc...
 *
 *  @author Jarrad Seers <jarrad@jarradseers.com>
 *  @created 07/08/2012 NZST
 */

/**
 *  Module dependencies.
 */

var fs = require('fs')
  , path = require('path')
  , util = require('util');

/**
 *  Module Globals.
 */

var title   = 'express-load';

/**
 *	Defaults.
 */

var extlist = ['.js', '.node', '.json', '.coffee', '.sjs'];
var checkext = true;

/**
 *  ExpressLoad constructor.
 */

var ExpressLoad = function(entity, options) {
  var options = options || {};
  options.checkext = options.checkext || checkext;
  options.extlist = options.extlist || extlist;
  options.logFormat = options.logFormat || '%s: %s';
  options.logger = options.logger || console;

  this.options = options;
  this.scripts = [];
  this.then(entity);

  return this;
};

/**
 *  Then method - creates a list of scripts in order.
 *
 *  @param location {String} Directory/file to load.
 *  @return this {Object} this ExpressLoad instance.
 */

ExpressLoad.prototype.then = function(location) {

  var entity = this.__getEntity(location);

  if (entity) {
    if (fs.statSync(entity).isDirectory()) {

      var dir = fs.readdirSync(entity);

      for (var file in dir) {

        var allow = true
          , script = path.join(entity, dir[file]);

        if (dir[file].charAt(0) === '.') {
          this.__log('Ignoring hidden entity: ' + dir[file], 'warn');

          allow = false;
        }

        if (allow && fs.statSync(script).isDirectory()) {
          this.then(path.join(location, dir[file]));
          allow = false;
        }

        for (var s in this.scripts) {
          if (this.scripts[s].fullpath === script) {
            allow = false;
          }
        }

        if(this.options.checkext && (this.options.extlist.indexOf(path.extname(dir[file]))==-1)) {
          this.__log('Ignoring file: extension not allowed ' + dir[file], 'warn');
          allow = false;
        }


        if (allow) {
          var name = getName(script);

          this.scripts.push({
            fullpath: script,
            location: location + path.sep + name,
            name: name
          });
        }

      }
    } else {
      var location = path.dirname(location)
        , name = getName(entity);

      this.scripts.push({
        fullpath: entity,
        location: location + path.sep + name,
        name: name
      });
    }
  }

  return this;
};

/**
 *  Create Namespace function - creates a namespace map.
 *
 *  @param parent {Object} Current register.
 *  @param parts {Array} Namespace location map.
 *  @return parent {Object} current namespace.
 */

function createNamespace(parent, parts) {

  var part = getName(parts.shift());

  if (!parent[part]) {
    parent[part] = {};
  }

  if (parts.length) {
    parent = createNamespace(parent[part], parts);
  }

  return parent;
}

/**
 *  Get Name - formats name into method-friendly version from script path.
 *
 *  @param script {String} Script location.
 *  @return name {String} Name of script.
 */

function getName(script) {
  var script = path.basename(script, path.extname(script)).replace(/[^a-zA-Z0-9]/g, '.')
    , parts = script.split('.')
    , name = parts.shift();

  if (parts.length) {
    for (var p in parts) {
      name += parts[p].charAt(0).toUpperCase() + parts[p].substr(1, parts[p].length);
    }
  }

  return name;
}

/**
 *  Get Entity Method - gets the formatted file or folder location.
 *
 *  @param location {String} Script location.
 *  @return entity {String} Normalised path.
 */

ExpressLoad.prototype.__getEntity = function(location) {
  if (typeof location !== 'string') {
    this.__log('File or Folder must be a string: ' + location, 'error');
    return false;
  }

  var entity = path.resolve(
    path.normalize(location)
  );

  if (!fs.existsSync(entity)) {
    this.__log('No such file or directory found at: ' + entity, 'error');
    return false;
  }

  return entity;
}

/**
 *  Into method, applies script to instance.
 *
 *  @param instance {Object} Express application instance.
 *  @return this {Object} this ExpressLoad instance.
 */

ExpressLoad.prototype.into = function(instance) {

  for (var s in this.scripts) {

    var script = this.scripts[s]
      , parts = script.location.split(path.sep)
      , maps = parts.slice(0)
      , ns = createNamespace(instance, parts)
      , mod = require(script.fullpath)
      , map = [];

    if (typeof mod === 'function') {
      mod = mod.apply(script, arguments);
    }

    ns[script.name] = mod;

    while (maps.length) {
      map.push(
        getName(
          maps.shift()
        )
      );
    }

    this.__log('Loaded *.' + map.join('.'));
  }

  return this;
};

/**
 *  Log method, Logs to console if verbose option is true.
 *
 *  @param message {String} Message to display.
 *  @param type {String} Type of message ('error', 'warn', 'info') defaults to info.
 *  @return this {Object} this ExpressLoad instance.
 */

ExpressLoad.prototype.__log = function(message, type) {
  if (this.options.verbose) {
    type = type || 'info';
    this.options.logger[type](this.options.logFormat, title, message);
  }
  return this;
};

/**
 *  Expose the ExpressLoad class.
 *
 *  @param entity {String} File or folder to load.
 *  @param app {Object} Express application instance.
 *  @return new ExpressLoad {Object} new ExpressLoad instance.
 */

module.exports = function(entity, options) {
  return new ExpressLoad(entity, options);
};
